[WIP] System stats plugin

alficles opened the pull request:
[alficles] Creating a pull request here, just so I can use the review functionality.

bryancall committed:
Limit the amount of memory that an HTTP/2 stream will allocate

scw00 committed:
move TRANSACT_RETURN to Transaction.cc

Commit:
bugfix for malformed chunked response will be cached under some terrible network circumstances,
clang-format code indent adjusted too!

SolidWallOfCode committed:
Move NetHandler initialization to be static and not per thread.
Remove update race conditions.
Related to issue #2761

scw00 committed:
add autest for post redirection

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+#define NET_STATS_DIR "/sys/class/net"
+#define STATISTICS_DIR "statistics"
+
+typedef struct {
```
It probably doesn't matter much here, but it's in general better to also give names to the structs directly. Otherwise, they cannot be forward-declared in other code. This struct almost certainly won't need to be, but it's something to think about in general. You can safely use the same name, too, if you wish.

ezelkow1 replied:
Removed struct

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+// with NET_STATS.infname.RX/TX.standard_net_stats field
+#define NET_STATS "plugin." PLUGIN_NAME ".net."
+
+#define NET_STATS_DIR "/sys/class/net"
```
Does `sysfs` have to be mounted at `/sys`? Or could it be mounted elsewhere. If there were a way to probe for it, it might increase portability.

ezelkow1 replied:
Found in the linux kernel doc rules for sysfs that it will always be at /sys, anything else is a complete system misconfiguration bug

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  TSMutex stat_creation_mutex;
+} config_t;
+
+// TBD: We used to keep track of config reload amts and times
```
Resolve this TBD. I would consider implementing it.

alficles replied:
If you don't implement it, remove it.

ezelkow1 replied:
removed, Rob mentioned we may not need but can always re-add it later. It is much more complicated to do with this plugin anyway since it is not intercepting the stats anymore so Id have to investigate

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+// We should only be grabbing these on a linux
+// or possibly BSD system. Others like OSX
+// do not have a proc or sysfs system
+#if defined(__linux__)
```
In general, it's much better to test for feature availability than to use `__linux__` to determine. This is an exercise in autoconf, which is great fun and never causes difficulty.

ezelkow1 replied:
autoconf'd things

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  }
+
+  s = TSfread(f, buffer, bufferSize);
+  if (s > 0)
```
`fread` can return less than the buffer requested if a read error or `eof` occurs. `eof` is the typical case, but you would not want to succeed on this function with a truncated file due to read error. (I realize this is very unlikely with `sysfs`, but safe programming is safer. :) ) Are the semantics for `TSfread` the same as `fread`?

alficles replied:
Likewise, if `s == bufferSize`, it means you haven't got the whole file. For most of these files, you should be reading basically a single integer. But if it's more than the buffer, that's probably an error as well.

ezelkow1 replied:
changed this function to return an int instead of char*. Char* made no sense anyway since it was just the buffer that was already being modified, so it returns -1 on some sort of read error

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+}
+
+static char *
+getFile(char *filename, char *buffer, int bufferSize)
```
`filename` is not modified, ought it be passed `const`?

ezelkow1 replied:
const'd all the things

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+}
+
+static void
+statSet(char *name, int value, TSMutex stat_creation_mutex)
```
`name` is not modified, ought it be passed `const`?

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+}
+
+static void
+setNetStat(TSMutex stat_creation_mutex, char *interface, char *entry, char *subdir)
```
I'm going to stop commenting on `char*` that should be const. You should do a pass to check for them, though.

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  memset(&data[0], 0, sizeof(data));
+
+  if ((interface == NULL) || (entry == NULL)) {
+    TSError("%s(): NULL subdir or entry", __FUNCTION__);
```
Most plugins define a "PluginError" macro that includes the name of the plugin in the error. This helps people narrow down what is failing.

alficles replied:
Also, file and line information is included by default if the user requests it in their configuration.

ezelkow1 replied:
From what Ive seen they all use TSError or TSDebug. TSDebug includes the name of the plugin but TSError does not. So I could change these to debugs but in the other plugins Ive see they seem to use error for these types of actual errors while debug is just more informational.

Im not sure what TSError may include by default in it's output though

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  }
+
+  while ((dent = readdir(srcdir)) != NULL) {
+    if (strcmp(dent->d_name, ".") == 0 || strcmp(dent->d_name, "..") == 0 || strcmp(dent->d_name, "lo") == 0) {
```
Should we just skip every "hidden" file that starts with a `.`?

ezelkow1 replied:
This doesnt do files, readdir only does dir names in the starting srcdir. But if there somehow happens to be a .devicedir inside the sysfs filesystem we should probably skip it anyway because that is strange, weird, and confusing

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+// We should only be grabbing these on a linux
+// or possibly BSD system. Others like OSX
+// do not have a proc or sysfs system
+#if defined(__linux__)
```
This is another place that autoconf should probably be used to determine if these stats should be collected. Alternately, it might be best to probe at initialization. Even a linux system may not have the statistics being reported, if I understand the settings correctly.

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+#include <unistd.h>
+
+#if defined (__linux__)
+#include <sys/sysinfo.h>
```
Autoconf can detect the presence of this header. If it's present, you should be able to configure this correctly, regardless of system.

ezelkow1 replied:
It looks like autoconf for ATS does already do a check for sysinfo, so the #ifdef check already works.  Question is do we want to just change these out for #ifdef HAVE_SYS_SYSINFO_H? Or just not build the plugin at all?

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  statSet(LOAD_AVG_FIVE_MIN, info.loads[1], stat_creation_mutex);
+  statSet(LOAD_AVG_TEN_MIN, info.loads[2], stat_creation_mutex);
+  statSet(CURRENT_PROCESSES, info.procs, stat_creation_mutex);
+  netStatsInfo(stat_creation_mutex);
```
Should netStatsInfo be unavailable if sysinfo is unavailable? I think you might be able to separate these.

ezelkow1 replied:
separated netstats out of the sysinfo stats AC switch. Of course they still probably won't get anything since sysfs most likely wont exist, but the option is there now

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  info.support_email = "dev@trafficserver.apache.org";
+
+  if (TSPluginRegister(&info) != TS_SUCCESS) {
+    TSError("[system_stats] Plugin registration failed");
```
Another good spot for a `PluginError` macro.

ezelkow1 replied:
changed the TSError usages to insert the plugin-name string

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+    TSError("[system_stats] Plugin registration failed");
+    return;
+  } else {
+    TSDebug(DEBUG_TAG, "Plugin registration succeeded");
```
Plugins also often use a PluginDebug macro that includes the plugin name with messages.

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+    TSDebug(DEBUG_TAG, "Plugin registration succeeded");
+  }
+
+  config                      = (config_t *)TSmalloc(sizeof(config_t));
```
This needs a comment to explain why it is never freed. I think that's correct, but a comment makes it explicit.

ezelkow1 replied:
I believe thats correct as well, but turns out we dont really need this anyway. The config_t was storing a mutex and a persistence type. We don't use the persistence type so it does not matter and we don't need the mutex either since one gets created during the TSContCreate and we can use the continuations built-in mutex

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  // Subsequent schedules done within the function will be for
+  // 5 seconds. One sec delay so hopefully we end up at the
+  // bottom of the stats list
+  TSContSchedule(stats_cont, 0, TS_THREAD_POOL_TASK);
```
This updates the stats asynchronously. Does that create the possibility that the endpoint could be hit prior to this being complete? This feels like a race condition to me.

ezelkow1 replied:
Its possible but in that case the stats just wont be there. The next time _stats gets hit then these system stats should be be there. If these were on their own endpoint I could see an issue but since they all get rolled in to _stats I dont think its that big of an issue

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+
+  // We want our first hit immediate to populate the stats,
+  // Subsequent schedules done within the function will be for
+  // 5 seconds. One sec delay so hopefully we end up at the
```
You say 1 sec delay, but then request 0 for the schedule time.

ezelkow1 replied:
fixed

alficles commented on plugins/experimental/system_stats/system_stats.c:
```diff
@@ -0,0 +1,278 @@
+  config->stat_creation_mutex = TSMutexCreate();
+
+  if (argc > 1) {
+    // config options if necessary
```
Either implement config options (stat-collection interval is a good one, I think), or leave this out. If the user provides unrecognized configuration arguments, that should be logged as an error, I think.

ezelkow1 replied:
removed, we can add as an arg later if we wish

alficles reviewed (commented):
@ezelkow1 This is my first round of feedback.

zwoop committed:
Enable WCCP building on the Github builds

ezelkow1 committed:
Create system_stats plugin. This will insert system information in to
the stats list on a five second interval.  Currently it pulls load
averages, number of processes, and most of the information from sysfs
for each available network device.

duke8253 committed:
make it python3 compatible

alficles committed:
Created URI Signing Plugin.

dragon512 committed:
Clean up tls test

Clean up some of the extension code
Add a general Build functions that gets flags from tsxs

bryancall committed:
Fixed broken build on OSX where st_mtim is not defined

bryancall committed:
Fixed build on Ubuntu, errors ignoring return value

bryancall committed:
Fixed another build issue on Ubuntu, errors ignoring return value

SolidWallOfCode committed:
Doc: Fix build warnings - SSL Session, Overridable Config.

persiaAziz committed:
Coverity: removing null pointer dereference

SolidWallOfCode committed:
Doc: Add Layer 4 / SNI Routing documentation.

dragon512 committed:
Rename prepare_plugin to PreparePlugin to follow naming convention

Add test for openssl version
Add condition test to some tests that need ALPN

bryancall committed:
coverity 1382795: Pointer to local outside scope

bryancall committed:
coverity 1382799, 1382796: Unchecked return value

bryancall committed:
coverity 1382722: Unchecked return value

persiaAziz-zz committed:
Delete ssl_SNI.config from tests/min_cfg

zwoop committed:
Add protocol metrics to traffic_logstats

zwoop committed:
Adds some basic documentation for traffic_logstats

shinrich committed:
Fix crash from HttpSM::setup_100_continue_transfer

masaori335 committed:
Suppress udpnet-service debug log

zwoop committed:
Reverts the changes from TS-3054

Also see the discussion on #2699, for a more complete fix.

SolidWallOfCode committed:
Fix potential crash from returning string_view from nullptr instead of default constructor.

dragon512 committed:
Fix a number of tests

change return code test as we can now shut doiwn better
Update min version of Autest
Change times on txn and double test to have time for metrics to be collection via contniuation. 
clean up test general-connection-failure
change kill timeout to 30 sec

zwoop committed:
Cleans up some build issue and unecessary casts

persiaAziz-zz committed:
overload ats_stringdup for string_view

zwoop committed:
Revert "bugfix for malformed chunked response will be cached under some terrible network circumstances,"

This reverts commit c2df5ba8d33b5774a35d5bc06289a7808149a337.

sfeltner-godaddy committed:
header_rewrite - Delete cookies if they are empty in rm-cookie operation.
